### PR TITLE
Adding Assessment Telemetry in SQL Migration

### PR DESCRIPTION
**Pull request by aasimkhan30**
Also fixed a blob container dropdown bug polluting console.

### CODE REVIEW

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 1]
```diff
@@ -196,12 +202,81 @@ export class MigrationStateModel implements Model, vscode.Disposable {
...
+						'sessionId': this._sessionId,
+						'hashedDatabaseName': hashString(d.name),
+						'compatibilityLevel': compatLevelHashMap.get(d.name)!,
+					},
+					{
+						'warningCount': d.issues.length,
+					}
+				);
+
+				d.issues.forEach(w => {
```
Comment (brian-harris) [inline extensions/sql-migration/src/models/stateMachine.ts:263]: sending telemetry for each database assessed and each issue seems a bit verbose.  is it possible to answer all the KPI's with an aggregate summary instead?

Comment (aasimkhan30) [reply extensions/sql-migration/src/models/stateMachine.ts:263]: I was under the assumption that we had to track every database getting assessed. This seemed like the best way to do it. Should I just add all the unique issues/warnings with their counts?

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 2]
```diff
@@ -196,12 +202,81 @@ export class MigrationStateModel implements Model, vscode.Disposable {
...
+				{
+					'issuesCount': this._assessmentResults.issues.length,
+					'duration': endTime - startTime,
+					'databaseCount': this._assessmentResults.databaseAssessments.length,
+					'startTime': startTime,
+					'endTime': endTime,
+				}
+			);
+
+			this._assessmentResults.issues.forEach(i => {
```
Comment (brian-harris) [inline extensions/sql-migration/src/models/stateMachine.ts:237]: do assessment exceptions show up in the issues list where the assessment tool was unable to assess a database / instance

Comment (aasimkhan30) [reply extensions/sql-migration/src/models/stateMachine.ts:237]: No, it does not. I will have to expose that from the backend

**Review by brian-harris — APPROVED** [review]

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 3]
```diff
@@ -196,12 +202,81 @@ export class MigrationStateModel implements Model, vscode.Disposable {
...
 		});
-
 		this._assessmentResults = {
 			issues: serverLevelAssessments,
 			databaseAssessments: databaseLevelAssessments
 		};
 
+		try {
+			const queryProvider = azdata.dataprotocol.getProvider<azdata.QueryProvider>(connectionProfile.providerId, azdata.DataProviderType.QueryProvider);
+			const query = 'SELECT name, compatibility_level FROM sys.databases; ';
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/models/stateMachine.ts:212]: I would strongly suggest putting all of this into an async function and not block this logic on it for perf reasons

File: extensions/sql-migration/src/models/stateMachine.ts (hunk 1 above)
Comment (Charles-Gagnon) [reply extensions/sql-migration/src/models/stateMachine.ts:263]: It'd be strongly preferred to aggregate counts whenever possible

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 4]
```diff
@@ -147,6 +152,7 @@ export class MigrationStateModel implements Model, vscode.Disposable {
 		this._databaseBackup = {} as DatabaseBackupModel;
 		this._databaseBackup.networkShare = {} as NetworkShare;
 		this._databaseBackup.blobs = [];
+		this._sessionId = uuidv4();
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/models/stateMachine.ts:155]: Just initialize it where you declare it, no reason to split that up.

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]
Waiting for confirmation on collection of OII/PII

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 5]
```diff
@@ -196,15 +201,61 @@ export class MigrationStateModel implements Model, vscode.Disposable {
...
+				{
+					'sessionId': this._sessionId,
+					'tenantId': this._azureAccount.properties.tenants[0].id,
+					'hashedServerName': hashString((connectionProfile.serverName))
+				},
+				{
+					'issuesCount': this._assessmentResults.issues.length,
+					'warningsCount': this._assessmentResults.databaseAssessments.reduce((count, d) => count + d.issues.length, 0),
+					'duration': endTime - startTime,
+					'databaseCount': this._assessmentResults.databaseAssessments.length,
```
Comment (brian-harris) [inline extensions/sql-migration/src/models/stateMachine.ts:234]: is this metric the number of databases on the server or number of assessed databases... could it be different if the assessment of a database fails?

**Comment by aasimkhan30** [comment]
@brian-harris  and @Charles-Gagnon, I have modified the events to answer the KPIs. Requesting a re-review of this PR.

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 6]
```diff
@@ -196,15 +201,104 @@ export class MigrationStateModel implements Model, vscode.Disposable {
...
+	private async generateAssessmentTelemetry(startTime: number, endTime: number, connectionProfile: azdata.connection.ConnectionProfile): Promise<void> {
+		try {
+			this._assessmentResults.issues.forEach(i => {
+				sendSqlMigrationActionEvent(
+					TelemetryViews.MigrationWizardTargetSelectionPage,
+					TelemetryAction.ServerAssessmentIssues,
+					{
+						'sessionId': this._sessionId,
+						'ruleId': i.ruleId
+
```
Comment (brian-harris) [inline extensions/sql-migration/src/models/stateMachine.ts:221]: can this extra inside line be removed

File: extensions/sql-migration/package.json [hunk 7]
```diff
@@ -89,6 +89,8 @@
   },
   "dependencies": {
     "@microsoft/ads-extension-telemetry": "^1.1.3",
+    "@types/uuid": "^8.3.0",
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/package.json:92]: `@types` packages should always go in devDependencies, they won't have any runtime components.

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 8]
```diff
@@ -196,15 +201,84 @@ export class MigrationStateModel implements Model, vscode.Disposable {
 				}
 			}
 		});
-
 		this._assessmentResults = {
 			issues: serverLevelAssessments,
 			databaseAssessments: databaseLevelAssessments
 		};
-
+		this.generateAssessmentTelemetry(startTime, endTime, connectionProfile);
```
[the remark this answers is not in the document]
Comment (Charles-Gagnon) [reply extensions/sql-migration/src/models/stateMachine.ts:208]: ping, usually non-awaited promises would be considered errors so it would be good to be explicit here

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 9]
```diff
@@ -575,6 +668,15 @@ export class MigrationStateModel implements Model, vscode.Disposable {
 
 	public async getBlobContainerValues(subscription: azureResource.AzureResourceSubscription, storageAccount: StorageAccount): Promise<azdata.CategoryValue[]> {
 		let blobContainerValues: azdata.CategoryValue[] = [];
+		if (!this._azureAccount || !subscription || !storageAccount) {
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/models/stateMachine.ts:671]: This doesn't seem related to this change, was this intentional?

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 10]
```diff
@@ -196,15 +201,103 @@ export class MigrationStateModel implements Model, vscode.Disposable {
...
+						'hashedDatabaseName': hashString(d.name),
+						'compatibilityLevel': compatLevelHashMap.get(d.name)!,
+					},
+					{
+						'warningsCount': d.issues.length
+					}
+				);
+
+				d.issues.forEach(i => {
+					if (databaseMap.has(i.ruleId)) {
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/models/stateMachine.ts:272]: A bit simpler would just be `databaseMap.set(i.ruleId, databaseMap.get(i.ruleId) ?? 0 + i.impactedObjects.length);` instead of the separate if/else checks

**Review by Charles-Gagnon — APPROVED** [review]
There's a bunch of unaddressed comments, but nothing blocking

**Comment by aasimkhan30** [comment]
@Charles-Gagnon , @brian-harris  added addtional events and modified the existing one.

File: extensions/sql-migration/src/dialog/migrationCutover/migrationCutoverDialogModel.ts [hunk 11]
```diff
@@ -34,18 +35,38 @@ export class MigrationCutoverDialogModel {
...
 					this._migration.subscription,
 					this.migrationStatus
 				);
+				sendSqlMigrationActionEvent(
+					TelemetryViews.MigrationCutoverDialog,
+					TelemetryAction.CutoverMigration,
+					{
+						'sessionId': this._migration.sessionId ?? '',
+						'migrationEndTime': new Date().toString()
+					},
```
Comment (brian-harris) [inline extensions/sql-migration/src/dialog/migrationCutover/migrationCutoverDialogModel.ts:66]: * migratonEndTime might by misleading, would 'cutoverStartTime' be more accurate.
* would it be safer to call the telemetry event before initiating the call in case it throws?

File: extensions/sql-migration/src/dialog/migrationCutover/migrationCutoverDialogModel.ts [hunk 12]
```diff
@@ -61,6 +82,15 @@ export class MigrationCutoverDialogModel {
 					this._migration.subscription,
 					this.migrationStatus
 				);
+				sendSqlMigrationActionEvent(
+					TelemetryViews.MigrationCutoverDialog,
+					TelemetryAction.CancelMigration,
+					{
+						'sessionId': this._migration.sessionId ?? '',
+						'migrationEndTime': new Date().toString()
```
Comment (brian-harris) [inline extensions/sql-migration/src/dialog/migrationCutover/migrationCutoverDialogModel.ts:90]: can sessionId be expected here and above, and not have to handle null

**Review by cssuh — APPROVED** [review]

File: extensions/sql-migration/package.json (hunk 7 above)
Comment (Charles-Gagnon) [reply extensions/sql-migration/package.json:154]: @aasimkhan30 ping

File: extensions/sql-migration/src/api/utils.ts [hunk 13]
```diff
@@ -157,6 +158,14 @@ export function findDropDownItemIndex(dropDown: DropDownComponent, value: string
 	return -1;
 }
 
+export function hashString(value: string): string {
+	return crypto.createHash('sha512').update(value).digest('hex');
+}
+
+export function getTime(): number {
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/api/utils.ts:165]: I don't see this actually being used anywhere

File: extensions/sql-migration/src/dialog/migrationCutover/migrationCutoverDialogModel.ts [hunk 14]
```diff
@@ -61,6 +82,15 @@ export class MigrationCutoverDialogModel {
 					this._migration.subscription,
 					this.migrationStatus
 				);
+				sendSqlMigrationActionEvent(
+					TelemetryViews.MigrationCutoverDialog,
+					TelemetryAction.CancelMigration,
+					{
+						'sessionId': this._migration.sessionId!,
+						'cutoverStartTime': new Date().toString()
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/dialog/migrationCutover/migrationCutoverDialogModel.ts:90]: This doesn't seem like the right event for sending cutoverStartTime - how do you know when the error actually occurred? Normally you'd capture the start time before starting the cutover and then include it in any relevant events afterwards.

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 15]
```diff
@@ -177,27 +184,160 @@ export class MigrationStateModel implements Model, vscode.Disposable {
...
+			issues: this._assessmentApiResponse.assessmentResult.items,
+			databaseAssessments: this._assessmentApiResponse.assessmentResult.databases.map(d => {
+				return {
+					name: d.name,
+					issues: d.items
+				};
+			})
 		};
-
+		this.generateAssessmentTelemetry();
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/models/stateMachine.ts:197]: Comment explaining that you're purposely not wanting to block on sending telemetry

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 16]
```diff
@@ -177,27 +184,160 @@ export class MigrationStateModel implements Model, vscode.Disposable {
...
 		return this._assessmentResults;
 	}
 
-	public findDatabaseAssessments(databaseName: string): mssql.SqlMigrationAssessmentResultItem[] | undefined {
-		return this._assessmentResults.databaseAssessments.find(databaseAsssessment => databaseAsssessment.name === databaseName)?.issues;
+	private async generateAssessmentTelemetry(): Promise<void> {
+		try {
+			this._assessmentResults.issues.forEach(i => {
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/models/stateMachine.ts:203]: How many issues are you expecting users to be getting? We want to be careful that we're not sending too many events all at once if we can avoid it.

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 17]
```diff
@@ -177,27 +184,160 @@ export class MigrationStateModel implements Model, vscode.Disposable {
...
+			const serverAssessmentErrorsMap: Map<number, number> = new Map();
+			this._assessmentApiResponse.assessmentResult.errors.forEach(e => {
+				if (serverAssessmentErrorsMap.has(e.errorId)) {
+					serverAssessmentErrorsMap.set(e.errorId, serverAssessmentErrorsMap.get(e.errorId)! + 1);
+				} else {
+					serverAssessmentErrorsMap.set(e.errorId, 1);
+				}
+			});
+
+			serverAssessmentErrorsMap.forEach((v, k) => {
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/models/stateMachine.ts:226]: Same here - what is the expected number of errors?

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 18]
```diff
@@ -177,27 +184,160 @@ export class MigrationStateModel implements Model, vscode.Disposable {
...
+					TelemetryAction.DatabaseAssessment,
+					{
+						'sessionId': this._sessionId,
+						'hashedDatabaseName': hashString(d.name),
+						'compatibilityLevel': d.compatibilityLevel
+					},
+					{
+						'warningsCount': d.items.length,
+						'errorsCount': d.errors.length,
+						'assessmentTime': d.assessmentTimeInMilliseconds,
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/models/stateMachine.ts:286]: assessmentTimeMs - always include units in measurement names

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 19]
```diff
@@ -177,27 +184,160 @@ export class MigrationStateModel implements Model, vscode.Disposable {
...
+						'numberOfBlockerIssues': d.sqlManagedInstanceTargetReadiness.numOfBlockerIssues,
+						'databaseSize': d.databaseSize
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/models/stateMachine.ts:288]: What are the units here?

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 20]
```diff
@@ -177,27 +184,160 @@ export class MigrationStateModel implements Model, vscode.Disposable {
...
+
+			databaseErrorsMap.forEach((v, k) => {
+				sendSqlMigrationActionEvent(
+					TelemetryViews.MigrationWizardTargetSelectionPage,
+					TelemetryAction.DatabaseAssessmentError,
+					{
+						'sessionId': this._sessionId
+					},
+					{
+						'errorId': k,
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/models/stateMachine.ts:332]: The errorId isn't really a measurement, this should probably be in the properties bucket

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 21]
```diff
@@ -177,27 +184,160 @@ export class MigrationStateModel implements Model, vscode.Disposable {
...
+					'endTime': this._assessmentApiResponse.endedTime,
+					'serverVersion': this._assessmentApiResponse.assessmentResult.serverVersion,
+					'serverEdition': this._assessmentApiResponse.assessmentResult.serverEdition,
+					'platform': this._assessmentApiResponse.assessmentResult.serverHostPlatform,
+					'engineEdition': this._assessmentApiResponse.assessmentResult.serverEngineEdition
+				},
+				{
+					'issuesCount': this._assessmentResults.issues.length,
+					'warningsCount': this._assessmentResults.databaseAssessments.reduce((count, d) => count + d.issues.length, 0),
+					'duration': endTime - startTime,
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/models/stateMachine.ts:260]: Units

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 22]
```diff
@@ -177,27 +184,160 @@ export class MigrationStateModel implements Model, vscode.Disposable {
...
+					'databaseCount': this._assessmentResults.databaseAssessments.length,
+					'serverHostCpuCount': this._assessmentApiResponse.assessmentResult.cpuCoreCount,
+					'serverHostPhysicalMemory': this._assessmentApiResponse.assessmentResult.physicalServerMemory,
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/models/stateMachine.ts:263]: units

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 23]
```diff
@@ -177,27 +184,160 @@ export class MigrationStateModel implements Model, vscode.Disposable {
...
+
+			serverAssessmentErrorsMap.forEach((v, k) => {
+				sendSqlMigrationActionEvent(
+					TelemetryViews.MigrationWizardTargetSelectionPage,
+					TelemetryAction.ServerAssessmentError,
+					{
+						'sessionId': this._sessionId
+					},
+					{
+						'errorId': k,
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/models/stateMachine.ts:234]: The errorId isn't really a measurement, this should probably be in the properties bucket

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 24]
```diff
@@ -704,14 +853,36 @@ export class MigrationStateModel implements Model, vscode.Disposable {
...
+							'sessionId': this._sessionId,
+							'migrationStartTime': new Date().toString(),
+							'targetDatabaseName': this._targetDatabaseNames[i],
+							'serverName': this._targetServerInstance.name,
+							'tenantId': this._azureAccount.properties.tenants[0].id,
+							'location': this._targetServerInstance.location,
+							'sqlMigrationService': this._sqlMigrationService.id,
+						},
+						{
+							'irRegistered': this._nodeNames.length > 0 ? 1 : 0
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/models/stateMachine.ts:873]: What does `irRegistered` mean? If this is just a boolean value it should probably go in the properties bucket

File: extensions/sql-migration/src/models/stateMachine.ts (hunk 16 above)
Comment (aasimkhan30) [reply extensions/sql-migration/src/models/stateMachine.ts:203]: Issues can range from a few (1-2) to many at a single time (100+)

Comment (aasimkhan30) [reply extensions/sql-migration/src/models/stateMachine.ts:203]: SInce we are just sending a map it should be 10-12 issues at the most

File: extensions/sql-migration/src/models/stateMachine.ts (hunk 17 above)
Comment (aasimkhan30) [reply extensions/sql-migration/src/models/stateMachine.ts:226]: 10-12 issues

File: extensions/sql-migration/src/models/stateMachine.ts (hunk 9 above)
Comment (aasimkhan30) [reply extensions/sql-migration/src/models/stateMachine.ts:723]: This was just a runtime error I caught

File: extensions/sql-migration/src/dialog/migrationCutover/migrationCutoverDialogModel.ts (hunk 14 above)
Comment (aasimkhan30) [reply extensions/sql-migration/src/dialog/migrationCutover/migrationCutoverDialogModel.ts]: The cutover start api to is asynchronous and returns a response immediately. However, I agree calculating the start time before calling the api will be more accurate.

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 25]
```diff
@@ -704,14 +855,37 @@ export class MigrationStateModel implements Model, vscode.Disposable {
...
+					sendSqlMigrationActionEvent(
+						TelemetryViews.MigrationWizardSummaryPage,
+						TelemetryAction.StartMigration,
+						{
+							'hashedServerName': hashString(this._assessmentApiResponse.assessmentResult.name),
+							'hashedDatabaseName': hashString(this._migrationDbs[i]),
+							'migrationMode': requestBody.properties.autoCutoverConfiguration ? 'online' : 'offline',
+							'sessionId': this._sessionId,
+							'migrationStartTime': new Date().toString(),
+							'targetDatabaseName': this._targetDatabaseNames[i],
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/models/stateMachine.ts:868]: The database and server names should be hashed
